src/Configuration.php

```diff
@@ -78,72 +89,125 @@ private function checkCredentials($username, $password)
+        if (null === $username || null === $password) {
+            // No change if credentials or null
+            return $this;
+        }
+
+        $new = $this->setValue('username', $username)
+            ->setValue('password', $password)
+            ->setValue('credentials', [$username, $password]);
+
+        // To keep BC
+        $new->username = $username;
+        $new->password = $password;
+        $new->credentials = [$username, $password];
+
+        return $new;
     }
 
+    /**
+     * @param string $interface
+     *
+     * @return Configuration
+     */
     public function bindToInterface($interface)
     {
-        return new self($this->getUsername(), $this->getPassword(), $interface, $this->getTimeout());
+        $new = $this->setValue('bind_to_interface', $interface);
+
+        // To keep BC
+        $new->bindtoInterface = $interface;
+
+        return $new;
     }
 
+    /**
+     * @param int $timeout
+     *
+     * @return Configuration
+     */
     public function withTimeout($timeout)
     {
-        return new self($this->getUsername(), $this->getPassword(), $this->getBindtoInterface(), $timeout);
+        $new = $this->setValue('timeout', $timeout);
+
+        // To keep BC
+        $new->timeout = $timeout;
+
+        return $new;
     }
 
     /**
      * @return array
+     *
+     * @deprecated Will be removed in 2.0. Use Configuration::getValue('credentials').
      */
     public function getCredentials()
     {
-        return $this->credentials;
+        return $this->getValue('credentials');
     }
 
     /**
      * @return string
+     *
+     * @deprecated Will be removed in 2.0. Use Configuration::getValue('bind_to_interface').
      */
     public function getBindtoInterface()
     {
-        return $this->bindtoInterface;
+        return $this->getValue('bind_to_interface');
     }
 
     /**
      * @return int
+     *
+     * @deprecated Will be removed in 2.0. Use Configuration::getValue('timeout').
      */
     public function getTimeout()
     {
-        return $this->timeout;
+        return $this->getValue('timeout');
     }
 
     /**
      * @return mixed
+     *
+     * @deprecated Will be removed in 2.0. Use Configuration::getValue('username').
      */
     public function getUsername()
     {
-        return $this->username;
+        return $this->getValue('username');
     }
 
     /**
      * @return mixed
+     *
+     * @deprecated Will be removed in 2.0. Use Configuration::getValue('password').
      */
     public function getPassword()
     {
-        return $this->password;
+        return $this->getValue('password');
     }
 
+    /**
+     * @param $mode
+     *
+     * @return Configuration
+     */
     public function withTLSMode($mode)
     {
-        $this->tlsMode = $mode;
-        return $this;
+        $new = $this->setValue('tls_mode', $mode);
+
+        // To keep BC
+        $new->tlsMode = $mode;
+
+        return $new;
```

Comment:
This is a BC break... Or maybe it is a bugfix. Depends on how you see it. 
`withTLSMode` was not immutable like all the other `with*` methods in this class. Now it is immutable. 

I would like to consider it a bugfix.